fix(core): throw better errors for potential circular references

crisbeto opened the pull request:
Currently circular references in user code manifest themselves with an error like `Cannot read properties of undefined (reading 'ɵcmp')`. This is a bit cryptic so these changes add an assertion mentioning circular references.

Relates to #65917.

JeanMeche reviewed (approved):
That's the kind of idea I had in the back of my mind too. 
As a follow-up we can probably create a dedicated doc page as this is an error that gets commonly reported.

JeanMeche reviewed (approved):
reviewed-for: public-api

JeanMeche commented on packages/core/src/render3/def_getters.ts:
```diff
@@ -52,13 +58,29 @@ export function getDirectiveDefOrThrow<T>(type: any): DirectiveDef<T> | never {
 export function getPipeDef<T>(type: any): PipeDef<T> | null {
+  if (typeof ngDevMode === 'undefined' || ngDevMode) {
+    assertTypeDefined(type, '@Pipe');
+  }
```
I'm wondering if we should also assert in prod mode too and only strip the error message in prod mode. 

What I've often seen reported is this issue happening only in prodmode because of the different bundler at play (vite in dev vs esbuild in prod).

crisbeto replied:
Updated it to also throw in prod mode.

thePunderWoman reviewed (approved):
reviewed-for: fw-general, public-api

alxhub reviewed (changes requested):
See @JoostK's comment.

crisbeto committed:
fix(core): throw better errors for potential circular references

Currently circular references in user code manifest themselves with an error like `Cannot read properties of undefined (reading 'ɵcmp')`. This is a bit cryptic so these changes add an assertion mentioning circular references.

Relates to #65917.

AndrewKushnir commented on packages/core/src/render3/def_getters.ts:
```diff
@@ -52,13 +54,26 @@ export function getDirectiveDefOrThrow<T>(type: any): DirectiveDef<T> | never {
+      RuntimeErrorCode.DEF_TYPE_UNDEFINED,
+      (typeof ngDevMode === 'undefined' || ngDevMode) &&
+        `Cannot read ${symbolType} metadata. This can indicate a runtime ` +
+          `circular dependency in your app that needs to be resolved.`,
```
Quick question: are those situations detectable? We do have some logic to handle circular dependencies in DI, should it be extended to catch more cases (in followup PRs)?

crisbeto replied:
I'm not sure they, at least not at runtime. Also my understanding is that some of them can happen after the code is processed by a bundler.

AndrewKushnir commented:
This PR was merged into the repository. The changes were merged into the following branches:

- main: 7be4ddef1ccb9a6b330b52b9bbdd174089755503
- 21.0.x: 4c8fb3631d58e22d693aba0b89243f2e9ecb0807
